File: app/schemas/git.py
"""
DevPilot Git Intelligence API Schemas.
"""

from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


class GitCommitInfoSchema(BaseModel):
    commit_hash: str = Field(..., description="Full SHA hash")
    short_hash: str = Field(..., description="Short 7-char SHA hash")
    author_name: str = Field(..., description="Author name")
    author_email: Optional[str] = Field(None, description="Author email")
    date: str = Field(..., description="Commit date timestamp")
    message: str = Field(..., description="Commit message")
    files_changed: List[str] = Field(default_factory=list, description="List of files modified")


class GitLastChangeResponse(BaseModel):
    symbol: str = Field(..., description="Target symbol or file path")
    commit: str = Field(..., description="Full commit SHA")
    short_hash: str = Field(..., description="Short 7-char commit SHA")
    author: str = Field(..., description="Author of the last change")
    date: str = Field(..., description="Commit timestamp")
    message: str = Field(..., description="Commit message")
    file: str = Field(..., description="Relative file path")
    line: Optional[int] = Field(None, description="Starting line number")
    end_line: Optional[int] = Field(None, description="Ending line number")


class GitHistoryResponse(BaseModel):
    symbol: str = Field(..., description="Target symbol or file path")
    file: str = Field(..., description="Resolved file path")
    line: Optional[int] = Field(None, description="Symbol start line if applicable")
    total_commits: int = Field(..., description="Total number of commits found")
    commits: List[GitCommitInfoSchema] = Field(default_factory=list, description="Chronological list of commits")


class GitBlameLineSchema(BaseModel):
    line_number: int = Field(..., description="Line number (1-indexed)")
    commit_hash: str = Field(..., description="Commit SHA")
    short_hash: str = Field(..., description="Short commit SHA")
    author: str = Field(..., description="Line author")
    date: str = Field(..., description="Commit timestamp")
    content: str = Field(..., description="Line source code content")


class GitBlameResponse(BaseModel):
    symbol: str = Field(..., description="Target symbol or file path")
    file: str = Field(..., description="Resolved file path")
    start_line: Optional[int] = Field(None, description="Start line number")
    end_line: Optional[int] = Field(None, description="End line number")
    total_lines: int = Field(..., description="Total blamed lines")
    primary_contributor: str = Field(..., description="Top contributor for this symbol/file")
    contributors: List[str] = Field(default_factory=list, description="List of distinct contributors")
    lines: List[GitBlameLineSchema] = Field(default_factory=list, description="Line-by-line blame records")


class GitCommitDetailResponse(BaseModel):
    commit_hash: str = Field(..., description="Full commit SHA")
    short_hash: str = Field(..., description="Short 7-char SHA")
    author_name: str = Field(..., description="Author name")
    author_email: Optional[str] = Field(None, description="Author email")
    date: str = Field(..., description="Commit timestamp")
    message: str = Field(..., description="Commit message")
    files_changed: List[str] = Field(default_factory=list, description="List of modified files")
    additions: int = Field(default=0, description="Number of lines added")
    deletions: int = Field(default=0, description="Number of lines deleted")
    diff_summary: str = Field(default="", description="Patch/diff summary")
    truncated: bool = Field(default=False, description="Whether diff was truncated due to size limit")

File: frontend/src/types/git.ts
export interface GitCommitDetail {
  commit_hash: string;
  short_hash: string;
  author_name?: string;
  author?: string;
  author_email?: string;
  email?: string;
  date: string;
  message: string;
  files_changed: string[];
  additions?: number;
  deletions?: number;
  diff_summary?: string;
  diff_patch?: string;
  stats?: {
    insertions: number;
    deletions: number;
    files: number;
  };
}

export interface GitLastChangeResponse {
  symbol?: string;
  symbol_or_file?: string;
  commit?: string;
  commit_hash?: string;
  short_hash: string;
  author?: string;
  author_name?: string;
  author_email?: string;
  date?: string;
  authored_date?: string;
  message?: string;
  commit_message?: string;
  file?: string;
  line?: number;
  end_line?: number;
}

export interface GitHistoryItem {
  commit_hash: string;
  short_hash: string;
  author?: string;
  author_name?: string;
  author_email?: string;
  date: string;
  message: string;
  files_changed?: string[];
}

export interface GitHistoryResponse {
  symbol?: string;
  symbol_or_file?: string;
  file?: string;
  line?: number;
  total_commits: number;
  commits: GitHistoryItem[];
}

export interface BlameLineItem {
  line_number: number;
  commit_hash: string;
  short_hash: string;
  author?: string;
  author_name?: string;
  date?: string;
  authored_date?: string;
  content?: string;
  line_content?: string;
}

export interface GitBlameResponse {
  symbol?: string;
  symbol_or_file?: string;
  file?: string;
  start_line?: number;
  end_line?: number;
  total_lines: number;
  primary_contributor?: string;
  contributors?: string[];
  lines: BlameLineItem[];
}

